pkg/services/ngalert/eval/datasource.go

```diff
@@ -0,0 +1,101 @@
+package eval
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/datasources"
+)
+
+// LastSeenDatasourceCache is a datasource cache, which on error, returns the last seen
+// version of the datasource.
+type LastSeenDatasourceCache struct {
+	cache       datasources.CacheService
+	log         log.Logger
+	datasources map[int64]*datasources.DataSource
+	uidsToIDs   map[string]int64
+	mu          sync.RWMutex
+}
+
+func NewLastSeenDatasourceCache(cache datasources.CacheService, log log.Logger) *LastSeenDatasourceCache {
+	return &LastSeenDatasourceCache{
+		cache:       cache,
+		log:         log,
+		datasources: make(map[int64]*datasources.DataSource),
+		uidsToIDs:   make(map[string]int64),
+	}
+}
+
+// GetDatasource returns the datasource with the ID. If there is an error on cache miss
+// then it returns the last seen version of the datasource.
+func (s *LastSeenDatasourceCache) GetDatasource(ctx context.Context, datasourceID int64, user *models.SignedInUser, skipCache bool) (*datasources.DataSource, error) {
+	datasource, err := s.cache.GetDatasource(ctx, datasourceID, user, skipCache)
+	if err != nil {
+		if errors.Is(err, datasources.ErrDataSourceNotFound) {
+			return nil, err
+		}
+		if datasource = s.getLastSeenDatasource(ctx, datasourceID, user); datasource != nil {
+			s.log.Info("failed to get datasource, using last seen datasource instead",
+				"datasourceID", datasourceID, "err", err)
+			return datasource, nil
+		}
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.datasources[datasource.Id] = datasource
+	return datasource, nil
+}
+
+func (s *LastSeenDatasourceCache) getLastSeenDatasource(_ context.Context, datasourceID int64, user *models.SignedInUser) *datasources.DataSource {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if v, ok := s.datasources[datasourceID]; ok {
```

Comment:
@yuri-tceretian do we need any more robust permission checking or has it already happened in an outer call before we fetch hit?

Reply:
I think we do not need a permission check on this level. If we use it in eval testing endpoint the check should be on the handler side

Reply:
I put the permission check in here just because its also in the datasource cache too.